module/zfs/zap_micro.c

```diff
@@ -1168,6 +1162,38 @@ zap_add(objset_t *os, uint64_t zapobj, const char *key,
 	}
 	ASSERT(zap == zn->zn_zap);
 	zap_name_free(zn);
+	return (err);
+}
+
+int
+zap_add(objset_t *os, uint64_t zapobj, const char *key,
+    int integer_size, uint64_t num_integers,
+    const void *val, dmu_tx_t *tx)
+{
+	zap_t *zap;
+	int err;
+
+	err = zap_lockdir(os, zapobj, tx, RW_WRITER, TRUE, TRUE, FTAG, &zap);
+	if (err != 0)
+		return (err);
+	err = zap_add_impl(zap, key, integer_size, num_integers, val, tx, FTAG);
+	if (zap != NULL)	/* may be NULL if fzap_add() failed */
```

Comment:
The comment is no longer true.  `zap` is passed by value, so there's no way for it to be changed.  You'll need to either pass a pointer to it (`&zap`), or have `zap_add_impl()` do the `zap_unlockdir()` for us.  Neither is that elegant, but I think the latter is less gross.